storagenode/storagenodedb/database.go

```diff
@@ -713,6 +777,61 @@ func (db *DB) Migration() *migrate.Migration {
 					)`,
 				},
 			},
+			{
+				DB:          db.deprecatedInfoDB,
+				Description: "Split into multiple sqlite databases",
+				Version:     22,
+				Action: migrate.Func(func(log *zap.Logger, _ migrate.DB, tx *sql.Tx) error {
+
+					// Migrate all the tables to new database files.
+					if err := db.migrateAndCloseDB(ctx, BandwidthDBName, "bandwidth_usage", "bandwidth_usage_rollups"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, OrdersDBName, "unsent_order", "order_archive_"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, PieceExpirationDBName, "piece_expirations"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, PieceInfoDBName, "pieceinfo_"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, PieceSpaceUsedDBName, "piece_space_used"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, ReputationDBName, "reputation"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, StorageUsageDBName, "storage_usage"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, UsedSerialsDBName, "used_serial_"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+					if err := db.migrateAndCloseDB(ctx, SatellitesDBName, "satellites", "satellite_exit_progress"); err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+
+					// Clean up the legacy database.
+					err := sqliteutil.KeepTables(ctx, db.rawDatabaseFromName(DeprecatedInfoDBName), "versions")
+					if err != nil {
+						return ErrDatabase.Wrap(err)
+					}
+
+					// Now close the deprecated db in order to free up unused
+					// disk space
+					if err := db.closeDatabase(DeprecatedInfoDBName); err != nil {
+						return ErrDatabase.Wrap(err)
```

Comment:
What would happen if this fails?

Reply:
The chain of errors will propagate until the entire storagenodedb will exit, which I believe will cause the entire process to exit.

Reply:
So when this fails, the next run would retry to run this migration, because the version doesn't get updated.

Reply:
correct

Reply:
The main thing I'm concerned about is that `if err := db.migrateAndCloseDB(ctx, BandwidthDBName, "bandwidth_usage", "bandwidth_usage_rollups"); err != nil {` when rerunning on an empty main database would destroy the second database.

Reply:
Simon and I are talking and realized that we failed to include the code that first deletes the destination db to make sure migration 22 is idempotent. Change incoming...

Reply:
I would also consider keeping the split to databases and clean up of the legacy databases in a separate migration. That way you can be sure that you don't delete any data before it has been migrated, and probably also slightly easier to reason about it.

Reply:
> I would also consider keeping the split to databases and clean up of the legacy databases in a separate migration.

Simon and I just came to the same conclusion while discussing ourselves. We'll drop the old tables in a separate migration step.

Reply:
We actually discovered more failure scenarios that can cause data loss during migration that we are currently implementing. For example, right now we issue each `DROP` SQL statement as a separate call and if any of those fail we on next startup we would re-run the migration with tables lost. So we are going to change that to be one atomic transaction. A half migrated destination isn't a big deal because we can delete the destination and re-try but we need to prevent situations where we have dropped data from the source.